app/Http/Controllers/TrainingDash.php

```diff
@@ -567,67 +573,366 @@ public function otsCancel($id) {
+                $percentSessionsChange = '-100';
+            } elseif ($retArr['sessionsPreviousMonth'] == 0) {
+                $percentSessionsChange = '+100';
+            } else {
+                $percentSessionsChange = round($retArr['sessionsPerMonth'] / $retArr['sessionsPreviousMonth'], 1);
+                if ($percentSessionsChange > 0) {
+                    $percentSessionsChange = '+' . $percentSessionsChange;
+                }
+            }
+            if ($retArr['sessionsCompletePerMonth'] == 0) {
+                $percentSessionsCompleteChange = '-100';
+            } elseif ($retArr['sessionsCompletePreviousMonth'] == 0) {
+                $percentSessionsCompleteChange = '+100';
+            } else {
+                $percentSessionsCompleteChange = round($retArr['sessionsCompletePerMonth'] / $retArr['sessionsCompletePreviousMonth'], 1);
+                if ($percentSessionsCompleteChange > 0) {
+                    $percentSessionsCompleteChange = '+' . $percentSessionsCompleteChange;
+                }
+            }
+            $retArr['taMonthlyReport'] = "In the Month of " . Carbon::createFromDate($retArr['date']['start_date'])->format('F') . ", ZTL has offered " . $retArr['sessionsPerMonth'] . " training sessions (" . $percentSessionsChange . "% change from " . Carbon::createFromDate($retArr['date']['start_date'])->subMonths(1)->format('F') . "). " . $retArr['sessionsCompletePerMonth'] . " sessions were completed (" . $percentSessionsCompleteChange . "%), with " . $retArr['sessionsPerMonthNoShow'] . " known no-shows. " . $trainingStaffBelowMins . " Training Staff members did not meet monthly minimums.";
+        }
+        return $retArr;
+    }
+
+    private function getAppointmentBatch(&$setmoreAppointments, $setmoreAccessToken, $from, $to, &$cursor) {
+        $cursorStr = '';
+        if (!is_null($cursor)) {
+            $cursorStr = '&cursor=' . $cursor;
+        }
+        $fromDate = Carbon::createFromDate($from)->format('d-m-Y');
+        $toDate = Carbon::createFromDate($to)->format('d-m-Y');
+        $client = new Client();
+        $res = $client->request(
+            'GET',
+            Config::get('setmore.endpoint') . '/bookingapi/appointments?startDate=' . $fromDate . '&endDate=' . $toDate . $cursorStr,
+            [
+                'headers' =>
+                [
+                    'Content-type' => "application/json",
+                    'Authorization' => "Bearer {$setmoreAccessToken}"
+                ]
+            ]
+        );
+        $setmoreResp = (string) $res->getBody();
+        $setmoreRespA = json_decode($setmoreResp, true);
+        if (isset($setmoreRespA['data']['appointments'])) {
+            $setmoreAppointments = array_merge($setmoreAppointments, $setmoreRespA['data']['appointments']);
+        }
+        if (isset($setmoreRespA['data']['cursor'])) {
+            $cursor = $setmoreRespA['data']['cursor'];
+            return true;
+        }
+        return false;
+    }
```

Comment:
See above note about moving the looping into this function

Reply:
Done